File: mask/app/src/main/java/com/example/masksearch/MainViewModel.java
package com.example.masksearch;

import android.location.Location;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.masksearch.model.Store;
import com.example.masksearch.model.StoreInfo;
import com.example.masksearch.repository.MaskService;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public class MainViewModel extends ViewModel {

    private static final String TAG = MainViewModel.class.getName();

    public MutableLiveData<List<Store>> itemLiveData = new MutableLiveData<>();
    public Location location;

    // Retrofit
    private Retrofit retrofit = new Retrofit.Builder()
            .baseUrl(MaskService.BASE_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .build();

    private MaskService service = retrofit.create(MaskService.class);

    private Call<StoreInfo> storeInfoCall = service.fetchStoreInfo();

    //Constructor
    public MainViewModel() {
        fetchStoreInfo();
    }

    public void fetchStoreInfo() {
        storeInfoCall.clone().enqueue(new Callback<StoreInfo>() {
            @Override
            public void onResponse(Call<StoreInfo> call, Response<StoreInfo> response) {
                Log.d(TAG, "on Response : Refresh");
                List<Store> items = response.body().getStores()
                        .stream()
                        .filter(item -> item.getRemainStat() != null)
                        .collect(Collectors.toList());

                itemLiveData.postValue(items);
            }

            @Override
            public void onFailure(Call<StoreInfo> call, Throwable t) {
                Log.e(TAG, "onFailure", t);
                itemLiveData.postValue(Collections.emptyList());
            }
        });
    }
}

File: mask/app/src/main/java/com/example/masksearch/MainViewModel.kt
package com.example.masksearch

import MaskService
import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.masksearch.model.Store
import com.google.android.gms.location.FusedLocationProviderClient
import kotlinx.coroutines.launch

class MainViewModel @ViewModelInject constructor(
    private val service: MaskService,
    private val fusedLocationProviderClient: FusedLocationProviderClient
) : ViewModel() {
    val itemLivaData = MutableLiveData<List<Store>>()
    val loadingLiveData = MutableLiveData<Boolean>()

    init {

        fetchStoreInfo()
    }

    @SuppressLint("MissingPermission")
    fun fetchStoreInfo() {
        loadingLiveData.value = true //  로딩 시작

        fusedLocationProviderClient.lastLocation
            .addOnSuccessListener { location ->
                viewModelScope.launch {
                    val storeInfo = service.fetchStoreInfo(location.latitude, location.longitude)
                    itemLivaData.value = storeInfo.stores.filter { store ->
                        store.remain_stat != null
                    }

                    loadingLiveData.value = false // 로딩 종료
                }
            }
            .addOnFailureListener {exception ->
                loadingLiveData.value = false // 로딩 종료
            }

    }


}